### ISSUE TITLE
Improve caching for manifest and lint

### ISSUE THREAD

**Issue by gsnedders** [web-platform-tests/wpt#11388]
Split out from https://github.com/web-platform-tests/wpt/issues/11356

One thing I've wanted to do for a long time is to massively improve caching for manifest and lint.

The status quo for manifest is we store a sha1 hash of the contents of each file and then compare it with a cached copy; this is expensive given it means we have to read every file to check if the cache is valid (this is especially notable on Windows due to https://github.com/web-platform-tests/wpt/issues/11386).

For lint we do all kinds of crazy magic to try and lint only a subset of files, based on git history.

We should instead:

 * use git blob IDs (in combination with the path where needed)
 * use file's mtime
 * maybe use hg node ids (or just the last revision, given I don't think hg has any public API or file format stability guarantees) for the sake of mozilla-central?

We need some combination of the above, given different things make sense in different cases (most obviously with untracked files but also the mozilla-central case).

We should absolutely be able to make manifest take almost no time at all when we have a cache (because we just need to verify that the tree still has the same blobs at the same paths) and make it possible for lint to always lint the entire tree mostly returning cached results (similarly in almost no time), though the simplest implementation would still to run all the `all_files_lints`.

Obviously any mtime-based caching would be filesystem specific, and hence not sharable. I'd like to suggest we add a local cache file, and make it optional to add a built-in cache to the generated manifest for the shared case. The local cache file could also easily store cached values for multiple revisions of a file, which would be useful.

**Comment by gsnedders** [comment]
To add:

This might be a good opportunity to do a larger refactoring where we can make things like `--work` and similar be consistent (e.g., `--work` should generate a manifest for the working tree; `--head` should generate one for HEAD rather than the current default of "files in HEAD but the copies currently in the working tree"; `--index` should generate one for the index).

**Comment by jgraham** [comment]
I have thoughts about this, which I'll add in the future, but for now @Cactusmachete who is likely to work on at least the manifest part of this.

**Comment by jgraham** [comment]
So in general this sounds pretty sensible. I agree that reading all the file contents to generate a hash is slow and would be better as a fallback path.

My concern with using the version control bits is that it's harder to make the cache distributed in the face of multiple environments and multiple version control systems. For example a requirement we have for gecko is that we are able to download a manifest into a tree that might be `hg` and might be `git` and immediately get good performance (i.e. we need to be able to download at least some level of caching). The file contents works well for this because it's independent of any other considerations like the particular VCS system or the specific history that we've built up.

One approach that could work is a multi-level cache. We would continue to store the file hashes in the manifest like today, but on top of that we would have a strictly local file mapping VCS blobs to file hashes. Then if the blob hasn't changed we wouldn't need to examine the file hash but (importantly) we could build the initial blob => file hash mapping with no more effort than it takes to do a zero-change update today.

I'm not exactly sure where you seem mtimes fitting in to this picture. Is the idea that you can use that to avoid having to ask the VCS when a file has changed since the current `HEAD`? 

For mercurial I'm pretty sure we can use `hg manifest --debug` to get the file revision hashes. It is *possible* that the format will change, but since it's only a cache I think we can live with that.

In terms of being pedantically correct about whether we are running on the source file or the file in VCS, I think that sounds nice, but it's worth remembering that the server will always serve the file from disk, so I don't think it makes much sense to spend lots of effort being perfect in that regard. Really the current difference isn't about worktree vs latest committed revision, but about whether or not to include untracked files. Maybe that's the difference we should make explicit.

**Comment by gsnedders** [comment]
>  For example a requirement we have for gecko is that we are able to download a manifest into a tree that might be `hg` and might be `git` and immediately get good performance (i.e. we need to be able to download at least some level of caching). The file contents works well for this because it's independent of any other considerations like the particular VCS system or the specific history that we've built up.

Is it acceptable in the Gecko case that the manifest that is downloaded must be generated within a Mercurial tree? At least in general it's done on TC in a Mercurial tree, right? Because then we can easily generate a cache file with both Mercurial last-revision-to-modify-file cache keys and git blob object id cache keys (because the blob object id depends on the file contents, so we can always compute it regardless of whether we're in a git tree, or even if we have git installed; well technically it also depends on the config given options like autocrlf and anything else that changes the content between the working tree and the blob so we can't guarantee a match but I expect almost all of the time it'll match?).

> I'm not exactly sure where you seem mtimes fitting in to this picture. Is the idea that you can use that to avoid having to ask the VCS when a file has changed since the current HEAD?

That's not the motivation: we shouldn't need to ask the VCS what's changed since the current HEAD for lint (we already don't for manifest) because we should be able to return the majority of lint results from cache so we can just always lint the whole tree.

The motivation here is that it deals with large numbers of local files (e.g., having the CSS built test suites in the working tree), because for performance we absolutely need them cached. I think for files that are only in the working tree we should deal with mtimes for rather than file content hashes, because it doesn't really make sense to put them in shared caches and mtimes will always be quicker to check than reading the entire contents.

> In terms of being pedantically correct about whether we are running on the source file or the file in VCS, I think that sounds nice, but it's worth remembering that the server will always serve the file from disk, so I don't think it makes much sense to spend lots of effort being perfect in that regard. Really the current difference isn't about worktree vs latest committed revision, but about whether or not to include untracked files. Maybe that's the difference we should make explicit.

I think this is useful for lint, at least, though I agree it might not be worthwhile for manifest.

For lint I think there's three different cases:

 * Checking your current working tree passes lint.
 * Checking the index passes lint pre-commit.
 * Checking a commit passes lint pre-push.

We probably also want options for the first to ignore ignored files (enabled by default) and to ignore untracked files (disabled by default).

The manifest the only argument to not use the working tree might be if you want to generate a shared manifest file from a commit with a dirty working tree, but I don't really see that happening at all often.

**Comment by gsnedders** [comment]
(Also mtimes mean that we even less often have to go and look at the VCS.)

**Comment by jgraham** [comment]
> Is it acceptable in the Gecko case that the manifest that is downloaded must be generated within a Mercurial tree?

Yes. If we can put both kinds of data in the cache without having git installed that will work well (I previosuly didn't realise that git blob ids didn't have any environment-specific metadata).

Are you imagining the cache being in the manifest itself, or in a seperate file? Given the mtime discussion below, I assume a separate file.

> That's not the motivation: we shouldn't need to ask the VCS what's changed since the current HEAD for lint (we already don't for manifest) because we should be able to return the majority of lint results from cache so we can just always lint the whole tree.

I guess that makes sense (using pure metadata as keys always worries me, but I guess there's a long history of using mtimes here). How do you imagine it working the first time you run this when you have (I assume) a VCS hash (assuming a tracked file), but no mtime (I'm assuming that we don't think they're portable). Do we then ask the VCS if the file changed, or do you imagine everyone having to do a full lint once?

> The motivation here is that it deals with large numbers of local files (e.g., having the CSS built test suites in the working tree), because for performance we absolutely need them cached

OK, but that seems like a really niche use case. Why are you doing that? :) But if the same work also helps with other common cases, it does seem worthwhile.

> For lint I think there's three different cases:

Yes, that seems reasonable if it's easy to implement.

**Comment by gsnedders** [comment]
> Are you imagining the cache being in the manifest itself, or in a seperate file? Given the mtime discussion below, I assume a separate file.

I'm imagining having an LRU cache probably in `$XDG_CACHE_HOME/web-platform-tests` (and equiv), and probably some option like `--shared` that adds an inline cache to MANFIEST.json giving shareable cache keys for which it is valid.

> How do you imagine it working the first time you run this when you have (I assume) a VCS hash (assuming a tracked file), but no mtime (I'm assuming that we don't think they're portable). Do we then ask the VCS if the file changed, or do you imagine everyone having to do a full lint once?

One option is that we distribute lint caches like we do for manifest, and use that as the seed data. (Of course, this *should* always just be "everything passes"!).

I'd say that making everyone run the lint once would be fine assuming #11386 is fixed, but in reality we'll need to invalidate the caches whenever tools/manifest or tools/lint change, which means it might be a bit too often?

**Comment by jgraham** [comment]
So I was talking to @Cactusmachete and I think a summary of the design would help us both. @gsnedders do you think the following makes sense / represents the ideas we discussed:

* In the manifest store git blob hashes rather than direct file hashes.
* For git, when files are updated, use `git ls-tree` to get the current blob hashes of each file, and use that to select the files to update.
* For hg, we can store the hg node id for each file at the time the manifest was last updated in a local (but sharable across gecko hg trees, I think) cache, and use that to figure out which files changed since the manifest was last updated.  
* We could also allow storing the commit for which the manifest was generated (with a clean working directory) and only examine files that changed since that commit. That works for the common case of having a manifest generated for a known commit and then downloaded.
* We could also locally store the `mtime` of each file the last time that the manifest was generated, and use that to avoid querying the VCS for information if no file mtimes changed (depending on the relative performance of these two operations).

**Comment by gsnedders** [comment]
@jgraham yes
